ui/endlessButton.js

```diff
@@ -0,0 +1,149 @@
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+const { Clutter, Gio, GObject, St } = imports.gi;
+
+const Main = imports.ui.main;
+const ViewSelector = imports.ui.viewSelector;
+
+const ExtensionUtils = imports.misc.extensionUtils;
+const PanelExtension = ExtensionUtils.getCurrentExtension();
+
+const OverviewWrapper = PanelExtension.imports.ui.overview;
+const SingleIconButton = PanelExtension.imports.ui.panelMenu.SingleIconButton;
+
+var EndlessButton = GObject.registerClass(
+class EndlessButton extends SingleIconButton {
+    _init() {
+        super._init(_('Endless Button'));
+        this.add_style_class_name('endless-button');
+        this.connect('style-changed', () => {
+            this.width = this.get_theme_node().get_length('width');
+            this.height = this.get_theme_node().get_length('height');
+        });
+        this.connect('notify::hover', this._onHoverChanged.bind(this));
+        this.connect('button-press-event', this._onButtonPressEvent.bind(this));
+        this.connect('button-release-event', this._onButtonReleaseEvent.bind(this));
+
+        let iconFile = Gio.File.new_for_uri(
+            `file://${PanelExtension.path}/data/icons/endless-button-symbolic.svg`);
+        this.setIcon(new Gio.FileIcon({ file: iconFile }));
+
+        this._setupTooltipText();
+    }
+
+    _setupTooltipText() {
+        this._label = new St.Label({ style_class: 'app-icon-hover-label' });
+
+        this._labelOffsetX = 0;
+        this._labelOffsetY = 0;
+        this._label.connect('style-changed', () => {
+            this._labelOffsetX = this._label.get_theme_node().get_length('-label-offset-x');
+            this._labelOffsetY = this._label.get_theme_node().get_length('-label-offset-y');
+        });
+
+        let pageChangedId = Main.overview.connect('page-changed', this._onOverviewPageChanged.bind(this));
+        let showingId = Main.overview.connect('showing', this._onOverviewShowing.bind(this));
+        let hidingId = Main.overview.connect('hiding', this._onOverviewHiding.bind(this));
+
+        this.connect('destroy', () => {
+            Main.overview.disconnect(pageChangedId);
+            Main.overview.disconnect(showingId);
+            Main.overview.disconnect(hidingId);
+        });
+
+        this._setHoverLabelText(true);
+    }
+
+    _updateHoverLabel(hiding) {
+        let viewSelector = Main.overview.viewSelector;
+        let showingDesktop = true;
+
+        if (!hiding &&
+            viewSelector &&
+            viewSelector.getActivePage() === ViewSelector.ViewPage.APPS)
+            showingDesktop = false;
```

Comment:
This could probably be simplified as

```javascript
const showindDesktop =
    hiding ||
    !viewSelector ||
    viewSelector.getActivePage() !== ViewSelector.ViewPage.APPS;
```

Reply:
Done.